src/player.cpp

```diff
@@ -11158,162 +11158,184 @@ void player::gunmod_add( item &gun, item &mod )
-    }
-
-    return HINT_GOOD;
+    std::vector<std::string> dummy;
+    const player *dummy_pl;
+    return time_to_read( it, dummy_pl, dummy ) == INT_MAX ? HINT_IFFY : HINT_GOOD;
 }
 
-bool player::read(int inventory_position)
+int player::time_to_read( const item &book, player const *&reader,
+                          std::vector<std::string> &reasons ) const
 {
-    // Find the object
-    item* it = &i_at(inventory_position);
-
-    if (it == NULL || it->is_null()) {
-        add_msg(m_info, _("You do not have that item."));
-        return false;
+    reader = nullptr;
+    if( !book.is_book() ) {
+        reasons.push_back( string_format( _( "Your %s is not good reading material." ),
+                                          book.tname().c_str() ) );
+        return INT_MAX;
     }
 
-    if (!it->is_book()) {
-        add_msg(m_info, _("Your %s is not good reading material."),
-        it->tname().c_str());
-        return false;
+    // Check for conditions that immediately disqualify the player from reading:
+    const vehicle *veh = g->m.veh_at( pos() );
+    if( veh != nullptr && veh->player_in_control( *this ) ) {
+        reasons.push_back( _( "It's a bad idea to read while driving!" ) );
+        return INT_MAX;
     }
-
-    //Check for NPCs to read for you, negates Illiterate and Far Sighted
-    int assistants = 0;
-    for( auto &elem : g->active_npc ) {
-        if (rl_dist( elem->pos(), g->u.pos() ) < PICKUP_RANGE && elem->is_friend()){
-            ///\EFFECT_INT_NPC allows NPCs to read harder books for you
-            if ((elem->int_cur+1) >= it->type->book->intel)
-                assistants++;
-        }
+    auto type = book.type->book.get();
+    if( type->fun <= 0 && !has_morale_to_read() && has_identified( book.typeId() ) ) {
+        // Low morale still permits skimming
+        reasons.push_back( _( "What's the point of studying?  (Your morale is too low!)" ) );
+        return INT_MAX;
     }
-
-    vehicle *veh = g->m.veh_at( pos() );
-    if( veh != nullptr && veh->player_in_control(*this) ) {
-        add_msg(m_info, _("It's a bad idea to read while driving!"));
-        return false;
+    const skill_id &skill = type->skill;
+    const auto &skill_level = get_skill_level( skill );
+    if( skill && skill_level < type->req ) {
+        reasons.push_back( string_format( _( "You don't know enough about %s to understand the jargon!" ),
+                                          skill.obj().name().c_str() ) );
+        return INT_MAX;
+    }
+
+    // Check for conditions that would disqualify us if no NPCs can read to us
+    if( fine_detail_vision_mod() > 4 ) {
+        reasons.push_back( _( "You can't see to read!" ) );
+    } else if( type->intel > 0 && has_trait( "ILLITERATE" ) ) {
+        reasons.push_back( _( "You're illiterate!" ) );
+    } else if( has_trait( "HYPEROPIC" ) && !is_wearing( "glasses_reading" ) &&
+               !is_wearing( "glasses_bifocal" ) && !has_effect( effect_contacts ) ) {
```

Comment:
There's `player::sight_impaired`, but that seems to be for seeing at a distance rather than close up. I searched the codebase for `hyperopic`, but didn't find anything.